test/msgpack.cpp

```diff
@@ -109,7 +109,11 @@ TEST_CASE(test_msgpack_object)
     auto buffer       = migraphx::to_msgpack(v);
     EXPECT(buffer == msgpack_buffer(std::map<std::string, double>{
                          {"one", 1.0}, {"three", 3.0}, {"two", 2.0}}));
-    EXPECT(migraphx::from_msgpack(buffer) == v);
+
+    // converted to vector in the following line because in value.cpp value constructor with
+    // unordered map is creating vector<value> with map items as vector elements
```

Comment:
Where are we using an unordered map?

Reply:
Created [buffer](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/a0894c2a541a58be81d3527b6e16b8f926c9061d/test/msgpack.cpp#L109) is a map. After its creation there is check if buffer is equal with buffer initialized with map [(here)](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/a0894c2a541a58be81d3527b6e16b8f926c9061d/test/msgpack.cpp#L110). 
In value definition inside [value.hpp](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/665a0cfb6e241aa11244d9be09d7415d46ee2fb5/src/include/migraphx/value.hpp) I haven't seen constructor for map, I found just constructor for unordered map, so because of that I used conversion to vector.

Reply:
> In value definition inside [value.hpp](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/665a0cfb6e241aa11244d9be09d7415d46ee2fb5/src/include/migraphx/value.hpp) I haven't seen constructor for map, I found just constructor for unordered map

Its not using the `std::unordered_map` constructor, its using the `std::initializer_list` constructor which will preserve the order of the list.

Reply:
I saw value constructor for std::initializer_list in [value.cpp](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/665a0cfb6e241aa11244d9be09d7415d46ee2fb5/src/value.cpp#L124). In this case (where size is not 2 and elements have key) it is calling `set_vector` function, so I think `to_vector()` function can be used for comparing if result values are equal.

Reply:
`to_vector` will only compare the values and not the keys. The `==` operator for value will compare the keys as well: `migraphx::from_msgpack(buffer) == v`.

Reply:
Yes, but `migraphx::from_msgpack(buffer)` and `v` are different types, so compare in `==` operator is not called. Because of that it should be converted to the same type.

Reply:
Yes, you are right, we can use `==` operator, but the problem here is that `migraphx::from_msgpack(buffer)` converts any numeric value to `uint64_t` (because of that return value is casted to `int` or `float`, for example in  `EXPECT(migraphx::from_msgpack(buffer).to<int>() == v.to<int>());`). 
If we cast input numeric value which is not uint8_t to uint8_t (for example in `migraphx::value v = {{"a", (uint64_t)1.0}, {"b", "abc"}};` inside `test_msgpack_object_class` or in `migraphx::value v = {{"one", (uint64_t) 1.0}, {"three", (uint64_t) 3.0}, {"two", (uint64_t) 2.0}};` inside `test_msgpack_object`, the test passes without cast `migraphx::from_msgpack(buffer)` and `v` because when comparing their numeric values they have the same type which is `uint8_t`. 

I think the best way here is to change  `migraphx::from_msgpack(buffer)` to cast properly numeric values which somehow it  doesn't cast and to see why is that happening.

Reply:
If `migraphx::from_msgpack(buffer)` function is fixed, I won't have need to cast output to `float` in the following way which i committed in this PR to (`EXPECT(migraphx::from_msgpack(buffer).to<float>() == v.to<float>());`).

Reply:
> but the problem here is that migraphx::from_msgpack(buffer) converts any numeric value to uint64_t

Where do you see this happening? `from_msgpack` converts the float32 or float64 to double here: https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/msgpack.cpp#L85.

Reply:
I see it in the debugger during comparing `EXPECT(migraphx::from_msgpack(buffer) == v);`. You can see debugger output in the following screenshot when debugging test_msgpack_float, here x is migraphx::from_msgpack(buffer) output and y is migraphx::value v which is double. If we cast v to be `uint64_t`, the test passes.
![image](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/assets/127323445/75352215-e5f7-4cc9-9d28-442edd7dff0a)

Reply:
My suggestion since we do want to keep comparing the keys and the types are different for the values that we should either change operator== to have a specialization that can properly compare types, call the compare function directly if that will allow us to compare types or create a compare function that can compare the keys and compare with an explicit type conversion (passed as a template parameter).

Reply:
So there is something wrong with either msgpack serialization or the value construction as it should be `float_type` as well.